quantus/helpers/model/tf_model.py

```diff
@@ -171,6 +171,71 @@ def get_random_layer_generator(self, order: str = "top_down", seed: int = 42):
             if order == "independent":
                 random_layer_model.set_weights(original_parameters)
             weights = layer.get_weights()
-            np.random.seed(seed=seed + 1)
+            np.random.seed(seed=seed + 1)  # noqa
             layer.set_weights([np.random.permutation(w) for w in weights])
             yield layer.name, random_layer_model
+
+    @cachedmethod(operator.attrgetter("cache"))
+    def _build_hidden_representation_model(
+        self, layer_names: Tuple, layer_indices: Tuple
+    ) -> Model:
+        # Instead of rebuilding model on each image, which is evaluated by metric, we cache it
+        if layer_names == () and layer_indices == ():
+            warn(
+                "quantus.TensorFlowModel.get_hidden_layers_representations(...) received `layer_names`=None and "
+                "`layer_indices`=None. This will force creation of tensorflow.keras.Model with outputs of each layer"
+                " from original model. This can be very computationally expensive."
+            )
+
+        def is_layer_of_interest(index: int, name: str) -> bool:
+            if layer_names == () and layer_indices == ():
+                return True
+            return index in layer_indices or name in layer_names
+
+        outputs_of_interest = []
+        for i, layer in enumerate(self.model.layers):
+            if is_layer_of_interest(i, layer.name):
+                outputs_of_interest.append(layer.output)
+
+        hidden_representation_model = Model(self.model.input, outputs_of_interest)
+        return hidden_representation_model
+
+    def get_hidden_representations(
+        self,
+        x: np.ndarray,
+        layer_names: Optional[List[str]] = None,
+        layer_indices: Optional[List[int]] = None,
+        **kwargs,
+    ) -> np.ndarray:
+        # List is not hashable, so we pass names + indices as tuples
+
+        num_layers = len(self.model.layers)
+
+        if layer_indices is None:
+            layer_indices = []
+
+        # Convert negative indices to positive
+        positive_layer_indices = [
+            i if i >= 0 else num_layers + i for i in layer_indices
+        ]
+        if layer_names is None:
+            layer_names = []
+
+        hidden_representation_model = self._build_hidden_representation_model(
```

Comment:
Please add comments in this function

Reply:
Is there a specific part which needs more clarification?